// @neo4j-ndl/react/lib/cjs/table/Table.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TableComponents = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
/**
 *
 *  Copyright (c) "Neo4j"
 *  Neo4j Sweden AB [http://neo4j.com]
 *
 *  This file is part of Neo4j.
 *
 *  Neo4j is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
const defaultImports_1 = require("../_common/defaultImports");
const helpers_1 = require("../helpers");
const react_1 = require("react");
const TableContext_1 = require("./TableContext");
const Components_1 = require("./Components");
/**
 *
 * Helping functions
 *
 */
exports.TableComponents = {
    ResizingBar: Components_1.ResizingBar,
    Header: Components_1.Header,
    Body: Components_1.Body,
    BodyCell: Components_1.BodyCell,
    BodyRow: Components_1.BodyRow,
    HeaderCell: Components_1.HeaderCell,
    ColumnControls: Components_1.ColumnControls,
    Navigation: Components_1.Navigation,
    TableResults: Components_1.TableResults,
    RowsPerPage: Components_1.RowsPerPage,
    Pagination: Components_1.Pagination,
    PaginationNumericButtons: Components_1.PaginationNumericButtons,
    PaginationNumericButton: Components_1.PaginationNumericButton,
    PaginationArrowButton: Components_1.PaginationArrowButton,
    LoadingPlaceholder: Components_1.LoadingPlaceholder,
    NoDataPlaceholder: Components_1.NoDataPlaceholder,
    SortingButton: Components_1.SortingButton,
    Scrollable: Components_1.Scrollable,
};
/**
 * @deprecated Use DataGrid component instead
 * Will be removed in Needle v3.0
 */
const Table = (0, helpers_1.forwardRef)(function Table(_a, ref) {
    var { className = '', style = {}, resizable = true, loading = false, focusableCells = false, tableProps, components } = _a, restProps = __rest(_a, ["className", "style", "resizable", "loading", "focusableCells", "tableProps", "components"]);
    const Components = (0, react_1.useMemo)(() => (Object.assign(Object.assign({}, exports.TableComponents), components)), [components]);
    const ScrollableContainer = (Components === null || Components === void 0 ? void 0 : Components.Scrollable)
        ? Components.Scrollable
        : Components_1.Scrollable;
    return ((0, jsx_runtime_1.jsx)(TableContext_1.TableContext.Provider, { value: {
            resizable,
            tableProps,
            loading,
            components: Components,
        }, children: (0, jsx_runtime_1.jsxs)("div", Object.assign({}, restProps, { className: (0, defaultImports_1.classNames)('ndl-table-root', className, {
                'ndl-table-focusable-cells': focusableCells,
            }), style: style, ref: ref, children: [(0, jsx_runtime_1.jsx)(ScrollableContainer, { children: (0, jsx_runtime_1.jsxs)("div", { className: "ndl-div-table", role: "table", "aria-busy": loading ? 'true' : 'false', children: [(Components === null || Components === void 0 ? void 0 : Components.Header) && (0, jsx_runtime_1.jsx)(Components.Header, {}), (Components === null || Components === void 0 ? void 0 : Components.Body) && (0, jsx_runtime_1.jsx)(Components.Body, {})] }) }), (Components === null || Components === void 0 ? void 0 : Components.Navigation) && (0, jsx_runtime_1.jsx)(Components.Navigation, {})] })) }));
});
/**
 * Easy Actions wrapper for each row
 */
const RowActions = (_a) => {
    var { className } = _a, restProps = __rest(_a, ["className"]);
    return ((0, jsx_runtime_1.jsx)("div", Object.assign({}, restProps, { className: (0, defaultImports_1.classNames)('ndl-table-row-actions', className) })));
};
// To work with forwardRef: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/34757#issuecomment-894053907
/**
 * @deprecated Use DataGrid component instead
 * Will be removed in Needle v3.0
 */
const TableNameSpace = Object.assign(Table, { RowActions });
exports.default = TableNameSpace;
