// @neo4j-ndl/react/lib/esm/drawer/Drawer.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
/**
 *
 *  Copyright (c) "Neo4j"
 *  Neo4j Sweden AB [http://neo4j.com]
 *
 *  This file is part of Neo4j.
 *
 *  Neo4j is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import { classNames } from '../_common/defaultImports';
import { useCallback } from 'react';
import { Resizable } from 're-resizable';
import { IconButton } from '../button';
import { XMarkIconOutline } from '../icons';
export const Drawer = (_a) => {
    var { children, className = '', expanded, onExpandedChange, position = 'left', type = 'overlay', isResizeable = false, resizeableProps, closeable = true } = _a, restProps = __rest(_a, ["children", "className", "expanded", "onExpandedChange", "position", "type", "isResizeable", "resizeableProps", "closeable"]);
    const setOnExpandedChange = useCallback(() => {
        if (onExpandedChange)
            onExpandedChange(!expanded);
    }, [expanded, onExpandedChange]);
    const classes = classNames('ndl-drawer', className, {
        'ndl-drawer-expanded': expanded,
        'ndl-drawer-left': position === 'left',
        'ndl-drawer-right': position === 'right',
        'ndl-drawer-overlay': type === 'overlay',
        'ndl-drawer-push': type === 'push',
    });
    const stylePosition = type === 'overlay' ? 'absolute' : 'relative';
    const Close = () => (_jsx(_Fragment, { children: closeable && (_jsx(IconButton, { className: "ndl-drawer-close-button", onClick: setOnExpandedChange, "aria-label": "close-drawer", size: "large", "data-testid": "ndl-drawer-close-button", title: "Close drawer", clean: true, children: _jsx(XMarkIconOutline, { className: "n-w-6 n-h-6" }) })) }));
    if (isResizeable) {
        return (_jsxs(Resizable, Object.assign({ defaultSize: {
                width: 'auto',
                height: '100%',
            } }, resizeableProps, { className: classes, style: Object.assign({ position: stylePosition }, resizeableProps === null || resizeableProps === void 0 ? void 0 : resizeableProps.style), boundsByDirection: true, bounds: "parent", enable: {
                top: false,
                right: position === 'left',
                bottom: false,
                left: position === 'right',
                topRight: false,
                bottomRight: false,
                bottomLeft: false,
                topLeft: false,
            }, children: [children, _jsx(Close, {})] })));
    }
    return (_jsxs("div", Object.assign({}, restProps, { className: classes, children: [children, _jsx(Close, {})] })));
};
const Header = (_a) => {
    var { children, className = '' } = _a, restProps = __rest(_a, ["children", "className"]);
    const classes = classNames('ndl-drawer-header', className);
    return (_jsx("h5", Object.assign({}, restProps, { className: classes, children: children })));
};
Drawer.Header = Header;
const Actions = (_a) => {
    var { children, className = '' } = _a, restProps = __rest(_a, ["children", "className"]);
    const classes = classNames('ndl-drawer-actions', className);
    return (_jsx("div", Object.assign({}, restProps, { className: classes, children: children })));
};
Drawer.Actions = Actions;
const Body = (_a) => {
    var { children, className = '' } = _a, restProps = __rest(_a, ["children", "className"]);
    const classes = classNames('ndl-drawer-body', className);
    return (_jsx("div", { className: "ndl-drawer-body-wrapper", children: _jsx("div", Object.assign({}, restProps, { className: classes, children: children })) }));
};
Drawer.Body = Body;
export default Drawer;
