// @neo4j-ndl/react/lib/cjs/layout/types.js
"use strict";
/**
 *
 *  Copyright (c) "Neo4j"
 *  Neo4j Sweden AB [http://neo4j.com]
 *
 *  This file is part of Neo4j.
 *
 *  Neo4j is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getRadius = exports.RadiusDefaults = exports.getSpacing = exports.SpacingDefaults = void 0;
const base_1 = require("@neo4j-ndl/base");
exports.SpacingDefaults = {
    padding: '4',
    paddingBlockStart: undefined,
    paddingBlockEnd: undefined,
    paddingInline: undefined,
    paddingInlineStart: undefined,
    paddingInlineEnd: undefined,
    gap: '4',
};
const getSpacing = (spacing, property) => {
    return spacing
        ? base_1.tokens.space[spacing || exports.SpacingDefaults[property]]
        : undefined;
};
exports.getSpacing = getSpacing;
exports.RadiusDefaults = {
    borderRadius: undefined,
};
const getRadius = (spacing, property) => {
    return spacing
        ? base_1.tokens.borderRadius[spacing || exports.RadiusDefaults[property]]
        : undefined;
};
exports.getRadius = getRadius;
