// @neo4j-ndl/base/lib/cjs/typescale.js
"use strict";
/**
 *
 *  Copyright (c) "Neo4j"
 *  Neo4j Sweden AB [http://neo4j.com]
 *
 *  This file is part of Neo4j.
 *
 *  Neo4j is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.typescale = void 0;
var tokens_esm_1 = require("./tokens/js/tokens-esm");
exports.typescale = {
    h1: [
        tokens_esm_1.tokens.font.size.h1,
        {
            letterSpacing: '-0.016rem',
            lineHeight: '3.75rem',
            fontWeight: tokens_esm_1.tokens.font.weight.bold,
        },
    ],
    h2: [
        tokens_esm_1.tokens.font.size.h2,
        {
            letterSpacing: '0',
            lineHeight: '3rem',
            fontWeight: tokens_esm_1.tokens.font.weight.bold,
        },
    ],
    h3: [
        tokens_esm_1.tokens.font.size.h3,
        {
            letterSpacing: '0.016rem',
            lineHeight: '2.5rem',
            fontWeight: tokens_esm_1.tokens.font.weight.bold,
        },
    ],
    h4: [
        tokens_esm_1.tokens.font.size.h4,
        {
            letterSpacing: '0.016rem',
            lineHeight: '2rem',
            fontWeight: tokens_esm_1.tokens.font.weight.bold,
        },
    ],
    h5: [
        tokens_esm_1.tokens.font.size.h5,
        {
            letterSpacing: '0.016rem',
            lineHeight: '1.75rem',
            fontWeight: tokens_esm_1.tokens.font.weight.bold,
        },
    ],
    h6: [
        tokens_esm_1.tokens.font.size.h6,
        {
            letterSpacing: '0.016rem',
            lineHeight: '1.5rem',
            fontWeight: tokens_esm_1.tokens.font.weight.bold,
        },
    ],
    'n-subheading-large': [
        // --font-size-subheading-large
        tokens_esm_1.tokens.font.size['subheading-large'],
        {
            letterSpacing: '0.016rem',
            lineHeight: '1.75rem',
            fontWeight: tokens_esm_1.tokens.font.weight.semibold,
        },
    ],
    'n-subheading-medium': [
        // --font-size-subheading-medium
        tokens_esm_1.tokens.font.size['subheading-medium'],
        {
            letterSpacing: '0.016rem',
            lineHeight: '1.5rem',
            fontWeight: tokens_esm_1.tokens.font.weight.semibold,
        },
    ],
    'n-subheading-small': [
        // --font-size-subheading-small
        tokens_esm_1.tokens.font.size['subheading-small'],
        {
            letterSpacing: '0',
            lineHeight: '1.25rem',
            fontWeight: tokens_esm_1.tokens.font.weight.semibold,
        },
    ],
    'n-body-large': [
        // --font-size-body-large
        tokens_esm_1.tokens.font.size['body-large'],
        {
            letterSpacing: '0',
            lineHeight: '1.5rem',
            fontWeight: tokens_esm_1.tokens.font.weight.normal,
        },
    ],
    'n-body-medium': [
        // --font-size-body-medium
        tokens_esm_1.tokens.font.size['body-medium'],
        {
            letterSpacing: '0',
            lineHeight: '1.25rem',
            fontWeight: tokens_esm_1.tokens.font.weight.normal,
        },
    ],
    'n-body-small': [
        // --font-size-body-small
        tokens_esm_1.tokens.font.size['body-small'],
        {
            letterSpacing: '0',
            lineHeight: '1.25rem',
            fontWeight: tokens_esm_1.tokens.font.weight.normal,
        },
    ],
    'n-code': [
        // --font-size-code
        tokens_esm_1.tokens.font.size.code,
        {
            letterSpacing: '0',
            lineHeight: '1.25rem',
            fontWeight: tokens_esm_1.tokens.font.weight.normal,
        },
    ],
    'n-label': [
        // --font-size-label
        tokens_esm_1.tokens.font.size.label,
        {
            letterSpacing: '0',
            lineHeight: '1.25rem',
            fontWeight: tokens_esm_1.tokens.font.weight.bold,
        },
    ],
};
