// @neo4j-ndl/react/lib/cjs/dropdown/Overrides.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomOption = exports.CustomPlaceholder = exports.CustomMultiValueRemove = exports.CustomClearIndication = exports.DropdownIndicatorCurrying = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
/**
 *
 *  Copyright (c) "Neo4j"
 *  Neo4j Sweden AB [http://neo4j.com]
 *
 *  This file is part of Neo4j.
 *
 *  Neo4j is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
const base_1 = require("@neo4j-ndl/base");
const defaultImports_1 = require("../_common/defaultImports");
const react_select_1 = require("react-select");
const icons_1 = require("../icons");
const DropdownIndicatorCurrying = ({ errorText, size, }) => function Component(props) {
    const { isDisabled, selectProps } = props;
    const classes = (0, defaultImports_1.classNames)('ndl-dropdown-icon', {
        'ndl-small': size === 'small',
        'ndl-medium': size === 'medium',
        'ndl-large': size === 'large',
        'ndl-disabled': isDisabled,
    });
    return ((0, jsx_runtime_1.jsx)(react_select_1.components.DropdownIndicator, Object.assign({}, props, { children: (0, jsx_runtime_1.jsxs)("div", { className: "n-flex n-flex-row n-gap-3", children: [(0, jsx_runtime_1.jsx)(icons_1.ChevronDownIconOutline, { className: classes, style: {
                        transform: selectProps.menuIsOpen
                            ? 'rotate(180deg)'
                            : 'rotate(0deg)',
                        transitionDuration: base_1.tokens.transitions.values.duration.quick,
                    } }), errorText && ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [(0, jsx_runtime_1.jsx)("div", { className: "ndl-indicator" }), (0, jsx_runtime_1.jsx)(icons_1.ExclamationCircleIconSolid, { className: `ndl-error-icon ${classes}`, role: "img", type: "solid" })] }))] }) })));
};
exports.DropdownIndicatorCurrying = DropdownIndicatorCurrying;
const CustomClearIndication = ({ errorText, size, }) => function Component(props) {
    return ((0, jsx_runtime_1.jsx)(react_select_1.components.ClearIndicator, Object.assign({}, props, { children: (0, jsx_runtime_1.jsxs)("div", { className: "n-flex n-flex-row n-gap-3", children: [(0, jsx_runtime_1.jsx)(icons_1.XMarkIconOutline, { className: "n-w-5 n-h-5 n-text-palette-neutral-text-weak n-cursor-pointer" }), (0, jsx_runtime_1.jsx)("div", { className: "ndl-indicator" })] }) })));
};
exports.CustomClearIndication = CustomClearIndication;
const CustomMultiValueRemove = ({ errorText, size, }) => function Component(props) {
    return ((0, jsx_runtime_1.jsx)(react_select_1.components.MultiValueRemove, Object.assign({}, props, { children: (0, jsx_runtime_1.jsx)(icons_1.XMarkIconOutline, { className: "n-text-palette-neutral-text-weak", style: {
                height: size === 'small' ? '16px' : '20px',
            } }) })));
};
exports.CustomMultiValueRemove = CustomMultiValueRemove;
const CustomPlaceholder = ({ errorText, size, }) => function Component(_a) {
    var { children } = _a, restProps = __rest(_a, ["children"]);
    return ((0, jsx_runtime_1.jsx)(react_select_1.components.Placeholder, Object.assign({}, restProps, { children: (0, jsx_runtime_1.jsx)("div", { className: "ndl-placeholder", children: children }) })));
};
exports.CustomPlaceholder = CustomPlaceholder;
const CustomOption = ({ errorText, size, }) => function Component(_a) {
    var { children } = _a, restProps = __rest(_a, ["children"]);
    const classes = (0, defaultImports_1.classNames)('ndl-dropdown-option', restProps.className, {
        'n-body-medium': size === 'small' || size === 'medium',
        'n-body-large': size === 'large',
    });
    return ((0, jsx_runtime_1.jsx)(react_select_1.components.Option, Object.assign({}, restProps, { className: classes, children: children })));
};
exports.CustomOption = CustomOption;
