// @neo4j-ndl/react/lib/cjs/dialog/Dialog.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Dialog = exports.DialogCloseReason = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
/**
 *
 *  Copyright (c) "Neo4j"
 *  Neo4j Sweden AB [http://neo4j.com]
 *
 *  This file is part of Neo4j.
 *
 *  Neo4j is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
const react_1 = __importStar(require("react"));
const defaultImports_1 = require("../_common/defaultImports");
const icons_1 = require("../icons");
const button_1 = require("../button");
const modal_1 = require("../modal");
// Should extend ModalCloseReason
var DialogCloseReason;
(function (DialogCloseReason) {
    DialogCloseReason["BACKDROP_CLICK"] = "backdropClick";
    DialogCloseReason["ESCAPE_KEY_DOWN"] = "escapeKeyDown";
    DialogCloseReason["CLOSE_BUTTON_CLICK"] = "closeButtonClick";
})(DialogCloseReason || (exports.DialogCloseReason = DialogCloseReason = {}));
const DialogContext = react_1.default.createContext(null);
const useDialogContext = () => {
    const value = (0, react_1.useContext)(DialogContext);
    if (!value) {
        throw new Error('Accessing `useDialogContext` outside of the context provider');
    }
    return value;
};
const convertCloseReason = (reason) => {
    const dialogReason = reason;
    if (Object.values(DialogCloseReason).includes(dialogReason)) {
        return dialogReason;
    }
    throw new Error(`Unknown reason '${reason}' received from popup`);
};
const typeIcons = {
    warning: (0, jsx_runtime_1.jsx)(icons_1.ExclamationTriangleIconOutline, { strokeWidth: "0.8" }),
    danger: (0, jsx_runtime_1.jsx)(icons_1.ExclamationCircleIconOutline, { strokeWidth: "0.8" }),
    info: (0, jsx_runtime_1.jsx)(icons_1.InformationCircleIconOutline, { strokeWidth: "0.8" }),
    success: (0, jsx_runtime_1.jsx)(icons_1.CheckCircleIconOutline, { strokeWidth: "0.8" }),
};
const Dialog = (_a) => {
    var _b;
    var { open, onClose, children, type, size, disableCloseButton = false } = _a, restProps = __rest(_a, ["open", "onClose", "children", "type", "size", "disableCloseButton"]);
    const handleCloseClick = (e) => {
        e.preventDefault();
        onClose && onClose(e, DialogCloseReason.CLOSE_BUTTON_CLICK);
    };
    const handleClose = (e, reason) => {
        onClose && onClose(e, reason ? convertCloseReason(reason) : undefined);
    };
    if (!open) {
        return null;
    }
    const classes = (0, defaultImports_1.classNames)('ndl-dialog', (_b = restProps.modalProps) === null || _b === void 0 ? void 0 : _b.className, {
        'ndl-with-icon': !!type,
        'ndl-with-close-button': !disableCloseButton,
    });
    const contentWrapperClasses = (0, defaultImports_1.classNames)('n-flex n-flex-nowrap', {
        'n-mt-10': !!type,
    });
    return ((0, jsx_runtime_1.jsx)(modal_1.Modal, Object.assign({}, restProps, { open: open, onClose: handleClose, modalProps: Object.assign(Object.assign({}, ((restProps === null || restProps === void 0 ? void 0 : restProps.modalProps) || {})), { className: classes, role: 'dialog' }), size: size, children: (0, jsx_runtime_1.jsxs)(DialogContext.Provider, { value: { type }, children: [!disableCloseButton && ((0, jsx_runtime_1.jsx)(button_1.IconButton, { clean: true, size: "large", className: "ndl-dialog-close", onClick: handleCloseClick, role: "button", "data-testid": "ndl-dialog-close", "aria-label": "close", children: (0, jsx_runtime_1.jsx)(icons_1.XMarkIconOutline, {}) })), (0, jsx_runtime_1.jsxs)("div", { className: contentWrapperClasses, children: [type && ((0, jsx_runtime_1.jsx)("div", { className: `ndl-dialog-type-icon ndl-${type}`, "data-testid": "ndl-dialog-type-icon", children: typeIcons[type] })), (0, jsx_runtime_1.jsx)("div", { className: "n-flex n-flex-col n-flex-1 n-w-full", children: children })] })] }) })));
};
exports.Dialog = Dialog;
const Actions = (_a) => {
    var { children, className = '' } = _a, rest = __rest(_a, ["children", "className"]);
    const { type } = useDialogContext();
    const classes = (0, defaultImports_1.classNames)('ndl-dialog-actions', className);
    let childrenToRender = children;
    // For danger dialogs, we want to convert primary buttons to danger buttons by default
    if (type === 'danger') {
        childrenToRender = react_1.default.Children.map(children, (element) => {
            if (!react_1.default.isValidElement(element))
                return;
            const { color } = element.props;
            if (!color || color === 'primary') {
                return react_1.default.cloneElement(element, Object.assign(Object.assign({}, element.props), { color: 'danger' }));
            }
            return element;
        });
    }
    return ((0, jsx_runtime_1.jsx)("div", Object.assign({}, rest, { className: classes, children: childrenToRender })));
};
exports.Dialog.Actions = Actions;
const Header = (_a) => {
    var { children, className = '' } = _a, rest = __rest(_a, ["children", "className"]);
    const classes = (0, defaultImports_1.classNames)('ndl-dialog-header', className);
    return ((0, jsx_runtime_1.jsx)("h4", Object.assign({}, rest, { className: classes, children: children })));
};
exports.Dialog.Header = Header;
const Subtitle = (_a) => {
    var { children, className } = _a, rest = __rest(_a, ["children", "className"]);
    const classes = (0, defaultImports_1.classNames)('ndl-dialog-subtitle n-body-large', className);
    return ((0, jsx_runtime_1.jsx)("div", Object.assign({}, rest, { className: classes, children: children })));
};
exports.Dialog.Subtitle = Subtitle;
const Description = (_a) => {
    var { children, className = '' } = _a, rest = __rest(_a, ["children", "className"]);
    const classes = (0, defaultImports_1.classNames)('ndl-dialog-description n-body-medium', className);
    return ((0, jsx_runtime_1.jsx)("div", Object.assign({}, rest, { className: classes, children: children })));
};
exports.Dialog.Description = Description;
const Content = (_a) => {
    var { children, className = '' } = _a, rest = __rest(_a, ["children", "className"]);
    const classes = (0, defaultImports_1.classNames)('ndl-dialog-content', className);
    return ((0, jsx_runtime_1.jsx)("div", Object.assign({}, rest, { className: classes, children: children })));
};
exports.Dialog.Content = Content;
